From 7101744ac38f21e20e5cd6a7b576e9fd779474d7 Mon Sep 17 00:00:00 2001
From: Sadrul Habib Chowdhury <sadrul@users.sourceforge.net>
Date: Tue, 21 Oct 2008 16:35:31 -0400
Subject: [PATCH] Fix memory corruption when unaliasing a command.

---
 src/process.c |   10 +++++-----
 1 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/src/process.c b/src/process.c
index 0024360..f5ac892 100644
--- a/src/process.c
+++ b/src/process.c
@@ -6562,7 +6562,7 @@ const char *name;
 {
   /* Find the previous alias */
   struct alias *cur  = g_aliases_list;
-  struct alias *prev = g_aliases_list;
+  struct alias **pcur = &g_aliases_list;
 
   while (cur != NULL)
     {
@@ -6573,7 +6573,7 @@ const char *name;
           int c;
 
           /* remove this one from the chain. */
-          prev->next = found->next;
+          *pcur = found->next;
 
           free(found->alias_name);
           free(found->alias_value);
@@ -6583,12 +6583,12 @@ const char *name;
                 free(found->alias_args[c]);
               free(found->alias_args);
             }
-          free(cur);
+          free(found);
 
           Msg(0, "alias %s removed", name);
-
+	  return;
         }
-      prev = cur;
+      pcur = &cur->next;
       cur = cur->next;
     }
 
-- 
1.5.6.2

