diff --git a/src/comm.c b/src/comm.c
index 0c56a0e..4c4ac0f 100644
--- a/src/comm.c
+++ b/src/comm.c
@@ -315,6 +315,7 @@ struct comm comms[RC_LAST + 1] =
   { "terminfo",		ARGS_23 },
   { "time",		CAN_QUERY|ARGS_01 },
   { "title",		CAN_QUERY|NEED_FORE|ARGS_01 },
+  { "tty",		CAN_QUERY|ARGS_0 },
   { "umask",		ARGS_1|ARGS_ORMORE },
   { "unbindall",	ARGS_0 },
   { "unsetenv",		ARGS_1 },
diff --git a/src/process.c b/src/process.c
index 866ed8e..d2ed9d6 100644
--- a/src/process.c
+++ b/src/process.c
@@ -2686,6 +2686,13 @@ int key;
       MakeTermcap((display == 0));
       debug("new termcap made\n");
       break;
+    case RC_TTY:
+      if(!args[0])
+      {
+        char buf[256];
+        OutputMsg(0, "%s", fore->w_tty);
+      }
+      break;
     case RC_ECHO:
       if (!msgok && (!rc_name || strcmp(rc_name, "-X")))
 	break;
